File: C++/15.三数之和.cpp

// 15. 三数之和
//
// 给你一个包含 n 个整数的数组 nums，判断 nums 中是否存在三个元素 a，b，c ，
// 使得 a + b + c = 0 ？请你找出所有和为 0 且不重复的三元组。
//
// 注意：答案中不可以包含重复的三元组。

#include <vector>

class Solution {
public:
    std::vector<std::vector<int>> threeSum(std::vector<int>& nums)
    {
        // std::sort(nums.begin(), nums.end());
        quick_sort(nums, 0, nums.size() - 1);

        std::vector<std::vector<int>> result;
        for (int i = 0; i < nums.size(); ++i) {
            if (nums[i] > 0) {
                // 剪枝
                break;
            } else if (i > 0 && nums[i] == nums[i - 1]) {
                // 跳过重复元素
                continue;
            }

            int l = i + 1, r = nums.size() - 1;
            while (l < r) {
                int sum = nums[i] + nums[l] + nums[r];
                if (sum > 0) {
                    --r;
                } else if (sum < 0) {
                    ++l;
                } else {
                    result.push_back({ nums[i], nums[l], nums[r] });

                    // 跳过重复元素
                    for (++l; l < r && nums[l] == nums[l - 1]; ++l) { }
                    for (--r; l < r && nums[r] == nums[r + 1]; --r) { }
                }
            }
        }

        return result;
    }

    void quick_sort(std::vector<int>& nums, int left, int right)
    {
        if (nums.empty() || left >= right) {
            return;
        }

        int l = left, r = right;
        while (l != r) {
            for (; l < r && nums[r] >= nums[left]; --r) { }
            for (; l < r && nums[l] <= nums[left]; ++l) { }
            if (l < r) {
                std::swap(nums[l], nums[r]);
            }
        }
        std::swap(nums[left], nums[l]);

        quick_sort(nums, left, l - 1);
        quick_sort(nums, r + 1, right);
    }
};

File: leetcode/01数组/15. 三数之和.go
package main

// 15. 三数之和

// 给你一个包含 n 个整数的数组nums，判断nums中是否存在三个元素 a，b，c ，
// 使得a + b + c = 0 ？请你找出所有和为 0 且不重复的三元组。
//
// 注意：答案中不可以包含重复的三元组。

// threeSum .
// 先固定一个值，接下来就是在剩余数字中寻找两数之和，但是需要保证结果不重复，
// 如果寻找两数之和时使用map的话，结果会重复；因此我们先将数组排序，在寻找两数之和时使用双指针从两边移动来找
// 排序 + 双指针
func threeSum(nums []int) [][]int {
	if nums == nil || len(nums) < 3 {
		return nil
	}

	quickSort(nums, 0, len(nums)-1)

	result := make([][]int, 0)
	for i := 0; i < len(nums); i++ {
		if nums[i] > 0 {
			// 剪枝 -- 若i指向大于0的数，则直接退出，因为数组已排序，i后面的数也肯定都是正数
			break
		} else if i > 0 && nums[i] == nums[i-1] {
			// 剪枝 -- 跳过重复元素
			continue
		}

		l, r := i+1, len(nums)-1
		for l < r {
			sum := nums[i] + nums[l] + nums[r]
			if sum > 0 {
				r--
			} else if sum < 0 {
				l++
			} else {
				// 记录结果
				result = append(result, []int{nums[i], nums[l], nums[r]})

				// 剪枝 -- 跳过重复元素
				for l++; l < r && nums[l] == nums[l-1]; l++ {
				}
				for r--; l < r && nums[r] == nums[r+1]; r-- {
				}
			}
		}
	}

	return result
}
